### PR TITLE
Add  Log Analytics 

### PR DESCRIPTION
**Pull request by MoeJw**
This PR closes issue #262

The changes in this PR are as follows:

* add a new resource to Farmer.

I have read the [contributing guidelines](CONTRIBUTING.md) and have completed the following:

* [x] **Tested my code** end-to-end against a live Azure subscription.
* [x] **Updated the documentation** in the docs folder for the affected changes.
* [x] **Written unit tests** against the modified code that I have made.
* [x] **Updated the [release notes](RELEASE_NOTES.md)** with a new entry for this PR.
* [x] **Checked the coding standards** outlined in the [contributions guide](CONTRIBUTING.md) and ensured my code adheres to them.

If I haven't completed any of the tasks above, I include the reasons why here:

### CODE REVIEW

File: src/Tests/Helpers.fs [hunk 1]
```diff
@@ -33,7 +33,7 @@ let convertResourceBuilder mapper (serializationSettings:Newtonsoft.Json.JsonSer
         |> fun json -> SafeJsonConvert.DeserializeObject(json, serializationSettings)
         |> Option.ofObj
     )
-
+//convertResourceBuilder ()(LogAnalytics)
```
Comment (isaacabraham) [inline src/Tests/Helpers.fs:36]: Remove commented out code.

File: src/Tests/LogAnalytics.fs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    test "can't create log analytics  with Sku eqaul to Standard and retentionInDays doesn't eqaul to 30 " {
+
+    
+    let e = logAnalytics { name "myFarmer"
+                           sku Standard
+                           retentionInDays 29
+                                 } :> IBuilder
+    Expect.throws (fun _ -> ( e.BuildResources Location.WestEurope |>ignore)) ""
+
```
Comment (isaacabraham) [inline src/Tests/LogAnalytics.fs:58]: Remove unnecessary whitespace / blank lines after each of the tests.

File: src/Tests/LogAnalytics.fs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    let c = logAnalytics { name "myFarmer"
+                           sku PerGB2018 
+                           retentionInDays 29
+                          } :> IBuilder
+    Expect.throws (fun _ -> ( c.BuildResources Location.WestEurope |>ignore)) "asdasdsadasdasdas" 
+    let d = logAnalytics {   name "myFarmer"
+                             sku PerNode 
+                             retentionInDays 29
+                            } :> IBuilder
+    Expect.throws (fun _ -> ( d.BuildResources Location.WestEurope |>ignore)) ""
```
Comment (isaacabraham) [inline src/Tests/LogAnalytics.fs:39]: Should have a space after `|>`.

Comment (isaacabraham) [inline src/Tests/LogAnalytics.fs:39]: No space after parenthesis e.g. `(c.BuildResources...`).

File: src/Tests/LogAnalytics.fs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    Expect.throws (fun _ -> ( e.BuildResources Location.WestEurope |>ignore)) ""
+    
+
+    }        
+
+    test "can't create log analytics  with Sku eqaul to Standard and retentionInDays doesn't eqaul to 365 " {
+    let f = logAnalytics { name "myFarmer"
+                           sku Premium
+                           retentionInDays 300 
+                                 } :> IBuilder
```
Comment (isaacabraham) [inline src/Tests/LogAnalytics.fs:66]: Align closing `}` with the opening one.

File: src/Farmer/Builders/Builders.LogAnalytics.fs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    /// Sets the name of the WorkSpace 
+    member _.Name(state: WorkSpaceconfig, name) = { state with Name = ResourceName name }
+
+    [<CustomOperation "sku">]
+    /// Sets the name of the SKU/Tier for the Log Analytics .
+    member _.Sku(state: WorkSpaceconfig,sku) = { state with Sku = sku }
+
+    [<CustomOperation "retentionInDays">]
+    ///The workspace data retention in days. -1 means Unlimited retention for the Unlimited Sku. 730 days is the maximum allowed for all other Skus /Standard and Premium pricing tiers which have fixed data retention of 30 and 365 days respectively..
+
```
Comment (isaacabraham) [inline src/Farmer/Builders/Builders.LogAnalytics.fs:55]: Remove whitespace line.

File: src/Farmer/Builders/Builders.LogAnalytics.fs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                                |_ ->  this.retentionInDays
+                        |None -> None
+                publicNetworkAccessForIngestion = this.publicNetworkAccessForIngestion
+                publicNetworkAccessForQuery = this.publicNetworkAccessForQuery
+                Tags=this.Tags
+                } ]
+type  WorkSpaceBuilder() =
+    /// Required - creates default "starting" values
+    member _.Yield _ =
+        {   Name = ResourceName.Empty
```
Comment (isaacabraham) [inline src/Farmer/Builders/Builders.LogAnalytics.fs:37]: Just one space between `{` and `Name` please.

File: src/Farmer/Builders/Builders.LogAnalytics.fs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+      Tags: Map<string,string>
+      }
+    interface IBuilder with
+        member this.DependencyName = this.Name
+        member this.BuildResources location =
+            [ { Name = this.Name
+                Location = location
+                Sku = this.Sku
+                retentionInDays = 
+                    match this.retentionInDays with
```
Comment (isaacabraham) [inline src/Farmer/Builders/Builders.LogAnalytics.fs:21]: reformat to standard F# guidelines e.g.

```fsharp
match this.retentionInDays with
| Standard when value <> 30 -> ...
| Premium when value <> 365 -> ...
```

etc.

File: src/Farmer/Builders/Builders.LogAnalytics.fs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                        |Some value ->
+                            match this.Sku with
+                                |Standard   when value<>30 -> failwithf "the retention Days for Standard must be 30."
+                                |Premium    when  value<>365 -> failwithf "the retention Days for Premium must be 365."
```
Comment (isaacabraham) [inline src/Farmer/Builders/Builders.LogAnalytics.fs:25]: Consider flipping the match to a tupled match to remove the need for `when` clauses e.g.

```fsharp
match this.Sku, this.RetentionInDays with
| Standard, Some 30 -> ...
| Standard, Some _ -> failwith "must be 30 days"
```

etc.

File: samples/scripts/loganalytics.fsx [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+open Farmer.Builders
+open Farmer.LogAnalytics
+
+let myRegistry = logAnalytics {
+    name "testmyFarmer2"
+    sku PerGB2018
+    retentionInDays 30
+    publicNetworkAccessForIngestion
+    publicNetworkAccessForQuery
+      }
```
Comment (isaacabraham) [inline samples/scripts/loganalytics.fsx:14]: Align closing brace per other samples.

File: samples/scripts/keyvault.fsx [hunk 10]
```diff
@@ -1,5 +1,5 @@
 #r @"./libs/Newtonsoft.Json.dll"
-#r @"../../src/Farmer/bin/Debug/netstandard2.0/Farmer.dll"
```
Comment (isaacabraham) [inline samples/scripts/keyvault.fsx:2]: In this updated path correct?

File: Farmer.sln [hunk 11]
```diff
@@ -7,28 +7,29 @@ Project("{6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705}") = "Farmer", "src\Farmer\Farmer
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "samples", "samples", "{2172A656-578A-48D4-81A2-126CBE1EF993}"
 EndProject
-Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "Tests", "src\Tests\Tests.fsproj", "{D68F5B81-6F18-4FC7-B9FD-2113EE7614CE}"
```
Comment (isaacabraham) [inline Farmer.sln:10]: Is there a reason for this file to change?

**Review by isaacabraham — CHANGES_REQUESTED** [review]
Mostly just some formatting changes. The only real code change I would suggest is around the pattern match.

**Comment by isaacabraham** [comment]
Thank you very much :-) Completed a review on this - mostly just aesthetic changes required.

**Comment by MoeJw** [comment]
Thanks for your attention. I look forward to doing more

**Comment by isaacabraham** [comment]
@MoeJw Have you tested this out end-to-end (as in, deployed a workspace using this)?

**Comment by isaacabraham** [comment]
Where are the rules coming from regarding retention period? Looking at the official ARM specs, it [simply says](https://docs.microsoft.com/en-us/azure/templates/microsoft.operationalinsights/workspaces#workspaceproperties-object:

> The workspace data retention in days. -1 means Unlimited retention for the Unlimited Sku. 730 days is the maximum allowed for all other Skus.

Terraform goes a [little further](https://www.terraform.io/docs/providers/azurerm/r/log_analytics_workspace.html):

> retention_in_days - (Optional) The workspace data retention in days. Possible values are either 7 (Free Tier only) or range between 30 and 730.

Where is the reference to these rules:

* Standard must be 30
* Premium must be 365
* Free must not be set

**Comment by isaacabraham** [comment]
Looking at the SKUs a little more, I am currently of the opinion that we should only support the new pricing PerGB2018 pricing tier plus capacity reservation options.

**Comment by MoeJw** [comment]
> Where are the rules coming from regarding retention period? Looking at the official ARM specs, it [simply says]

Yes, that's what I did first I tried to follow their specs and tested it by deploying it on Azure. 
I tried different scenarios but there was something wrong with it

image

I found out later about this article describing  the azure Manage usage and costs with Azure Monitor Logs
[https://docs.microsoft.com/en-us/azure/azure-monitor/platform/manage-cost-storage#change-the-data-retention-period](url)

>Workspaces in the Standalone or Per Node pricing tiers have user-configurable retention from 30 to 730 days.

> the original Standard and Premium pricing tiers which have fixed data retention of 30 and 365 days respectively.

I tested these as described on Azure  and it worked
- Standard 30
- Premium  365
- Standalone, Per Node and PerGB 30-730

> If you specify a pricing tier of Free, then remove the retentionInDays element.
[https://docs.microsoft.com/en-us/azure/azure-monitor/samples/resource-manager-workspace](url)

but here is the thing Standard, Premium, Standalone, and Per Node (OMS)  these are all legacy pricing tiers
so if you try to create LogAnalytics on Azure portal the only option provided is PerGB
however, you can  still use these legacy tiers when you deploy them as an arm template (:

**here is an example when you try to create log analytics on Azure portal** 

image

I was confused a little bit about it so I decided that I will leave it for the user to choose. but my guess now we should only support the new pricing PerGB2018

**Comment by isaacabraham** [comment]
@MoeJw I've updated the PR with a set of changes I made yesterday. Most of them were around code style and clean up - I'd suggest looking through them as it provides a good example of the coding style and standards for the project for future reference.

I also made some changes to the way we model SDK and Retention Period - instead of leaving them as two separate fields, we now model the relationship between them, which removes the need for much of the validation because it's now impossible to create e.g. Standard with a Retention Period at all.

I tested it out for both the Per GB and at least one legacy option. The UI in the portal doesn't offer the legacy options any more, unless you have created a resource using a legacy SKU. For that reason I think we should avoid providing them; if we end up just going with the new PerGB model the model can be simplified further.

I also noticed some undocumented features on the ARM template that are available in the UI. There's also no documentation that I could find on the Capacity Reservation options. @alex-frankel do you have any idea if there is any up-to-date docs on the Log Analytics resource - [this link](https://docs.microsoft.com/en-us/azure/templates/microsoft.operationalinsights/workspaces) is pretty out of date:

* It references something called the Unlimited Sku which doesn't even exist.
* No documentation on how to specify capacity reservation
* Daily Cap is available in the UI but not mentioned here. Exporting a template shows the following as a property:

```json
"workspaceCapping": {
    "dailyQuotaGb": 1
}
```

but this is not mentioned in the specs.

**Review by MoeJw — COMMENTED** [review]
can we use a tool like fantomas for formating?

**Comment by MoeJw** [comment]
> @MoeJw I've updated the PR with a set of changes I made yesterday. Most of them were around code style and clean up - I'd suggest looking through them as it provides a good example of the coding style and standards for the project for future reference.

Thank you so much, I will try to follow the standards next time sorry

**Comment by isaacabraham** [comment]
@MoeJw no worries - I'm really grateful that you've taken the time to get involved, it's a new codebase etc. and I'm happy to help out :-)

**Comment by isaacabraham** [comment]
There's actually an issue about fantomas that's already been opened - I should really look into that.

**Comment by MoeJw** [comment]
any Further changes requested for this PR  ?

**Comment by isaacabraham** [comment]
@MoeJw I think it's now at a state where it's ready to go. I'm just waiting to hear from Alex regarding the questions above; the only change we might make is removing the SKU type completely and just leaving it as the PerGB one, which would mean removing the sdk keyword and re-introducing the retention_period one.

**Comment by alex-frankel** [comment]
Sent an email that you are cc'd on @isaacabraham 

Unfortunately, I don't know the answers. That being said, my position would be is if it is *possible* via the PUT API, it should be supported in Farmer because someone may need those properties -- even if they are not documented. I can understand wanting to hide them though.

**Comment by isaacabraham** [comment]
@alex-frankel thanks again for looping me in.

The daily cap is something that's exposed in the UI and *appears* to get emitted by the portal-generated template. With the other stuff, it just looks like the documentation is incomplete - the docs refer to the capacity reservation but don't say anything about how to actually use it :-)

**Review by isaacabraham — APPROVED** [review]

**Comment by isaacabraham** [comment]
@MoeJw I can see you've checked that you've updated the release notes but I can't find the update there?

**Comment by isaacabraham** [comment]
Also I've updated the docs with my changes but I think that there are a few elements still missing like Tags. If you could just add those, I can then merge this in.

**Comment by MoeJw** [comment]
@isaacabraham  yeah I think the only thing left in docs is Tags.

**Pull request merged**